File: backend/app/Models/Schedule.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Builder;

class Schedule extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'user_id', 'title', 'category', 'day_of_week',
        'start_time', 'end_time', 'room', 'lecturer',
        'recurrence', 'start_date', 'end_date', 'color_tag', 'is_active', 'notes',
    ];

    protected function casts(): array
    {
        return [
            'day_of_week' => 'integer',
            'start_date'  => 'date',
            'end_date'    => 'date',
            'is_active'   => 'boolean',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function scopeToday(Builder $query): Builder
    {
        $today = now()->dayOfWeek; // Carbon: 0=Sun
        return $query->where('day_of_week', $today)
            ->where('is_active', true)
            ->where('start_date', '<=', today())
            ->where(function ($q) {
                $q->whereNull('end_date')->orWhere('end_date', '>=', today());
            })
            ->orderBy('start_time');
    }

    public function scopeThisWeek(Builder $query): Builder
    {
        $days = [];
        for ($i = 0; $i < 7; $i++) {
            $days[] = now()->addDays($i)->dayOfWeek;
        }
        return $query->whereIn('day_of_week', array_unique($days))
            ->where('is_active', true)
            ->where('start_date', '<=', now()->addDays(7))
            ->where(function ($q) {
                $q->whereNull('end_date')->orWhere('end_date', '>=', today());
            });
    }
}

File: mobile/lib/features/schedule/domain/entities/schedule.dart
// lib/features/schedule/domain/entities/schedule.dart

class ScheduleEntity {
  final int id;
  final String title;
  final String category;
  final int dayOfWeek;
  final String dayName;
  final String startTime;
  final String endTime;
  final int durationMinutes;
  final String? room;
  final String? lecturer;
  final String recurrence;
  final DateTime startDate;
  final DateTime? endDate;
  final String colorTag;
  final bool isActive;
  final bool isToday;
  final String? notes;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const ScheduleEntity({
    required this.id,
    required this.title,
    required this.category,
    required this.dayOfWeek,
    required this.dayName,
    required this.startTime,
    required this.endTime,
    required this.durationMinutes,
    this.room,
    this.lecturer,
    required this.recurrence,
    required this.startDate,
    this.endDate,
    required this.colorTag,
    required this.isActive,
    required this.isToday,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  ScheduleEntity copyWith({
    int? id,
    String? title,
    String? category,
    int? dayOfWeek,
    String? dayName,
    String? startTime,
    String? endTime,
    int? durationMinutes,
    String? room,
    String? lecturer,
    String? recurrence,
    DateTime? startDate,
    DateTime? endDate,
    String? colorTag,
    bool? isActive,
    bool? isToday,
    String? notes,
  }) =>
      ScheduleEntity(
        id: id ?? this.id,
        title: title ?? this.title,
        category: category ?? this.category,
        dayOfWeek: dayOfWeek ?? this.dayOfWeek,
        dayName: dayName ?? this.dayName,
        startTime: startTime ?? this.startTime,
        endTime: endTime ?? this.endTime,
        durationMinutes: durationMinutes ?? this.durationMinutes,
        room: room ?? this.room,
        lecturer: lecturer ?? this.lecturer,
        recurrence: recurrence ?? this.recurrence,
        startDate: startDate ?? this.startDate,
        endDate: endDate ?? this.endDate,
        colorTag: colorTag ?? this.colorTag,
        isActive: isActive ?? this.isActive,
        isToday: isToday ?? this.isToday,
        notes: notes ?? this.notes,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}
